tests/test_polynomials_operations.py

```diff
@@ -258,7 +259,7 @@ def test_mul_zero_result_zero(self):
         p3 *= 0.0
         p4 *= 0j
 
-        self._assert_polynomials_are_the_same(z0, result1)
+        self._assert_polynomials_are_the_same(z1, result1)
         self._assert_polynomials_are_the_same(z0, result2)
```

Comment:
With the changes so far:
`Polynomial(...) * ZeroPolynomial()` ==> `Polynomial()`
`ZeroPolynomial() * Polynomial(...)` ==> `ZeroPolynomial()`
Didn't we want `ZeroPolynomial() * Polynomial(...)` to produce `Polynomial()` as well?

Reply:
It does seem better to return the most permissive class - I've changed the behaviour to reflect this, and added tests to ensure this is maintained.